### DISCUSSION TITLE
Helm chart versions in values.yaml

### DISCUSSION THREAD

**Discussion by fuero** [renovatebot/renovate#29764] [Request Help]
### What would you like help with?

I would like help with my configuration

### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

Self-hosted Gitlab, renovatebot 37.412.2

### Please tell us more about your question or problem

I have a helm chart that creates Flux `HelmRelease` resources, using config blocks in `values.yaml` like this:

```yaml
myapp:
  # renovate: depName=myapp
  chartVersion: 1.2.3
```

This works fine with this config:
```json
{
"customManagers": [
    {
      "customType": "regex",
      "fileMatch": [
        "chart/values[.]yaml$"
      ],
      "matchStrings": [
        "\\s*#\\s+renovate:\\s+(?:datasource=(?<datasource>\\S+)\\s+)?depName=(?<depName>\\S+)(?:\\s+registryUrl=(?<registryUrl>\\S+))?(?:\\s+versioning=(?<versioning>\\S+))?(?:\\s+extractVersion=(?<extractVersion>\\S+))?\\s+chartVersion:\\s+(?<currentValue>\\S+)"
      ],
      "extractVersionTemplate": "{{#if extractVersion}}{{{extractVersion}}}{{else}}^(?<version>.+)${{/if}}",
      "datasourceTemplate": "{{#if datasource}}{{{datasource}}}{{else}}docker{{/if}}",
      "versioningTemplate": "{{#if versioning}}{{{versioning}}}{{else}}loose{{/if}}",
      "registryUrlTemplate": "{{#if registryUrl}}{{{registryUrl}}}{{else}}https://internal.registry/{{/if}}"
    }
  ]
}
```

But Helm supports version ranges, with which this config has trouble:

```yaml
myapp:
  # renovate: depName=myapp
  chartVersion: =1.2.3
```

yielding:

```
DEBUG: HEAD https://internal.registry/v2/myapp/manifests/=1.2.3 = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=404 retryCount=0, duration=56) (repository=local)r
[...]
DEBUG: Could not determine new digest for update. (repository=local)
       "packageName": "myapp",
       "currentValue": "=1.2.3",
       "datasource": "docker",
       "newValue": "=1.2.3",
       "bucket": undefined
```

Running helm itself like this works fine though:

```
helm pull "oci://internal.registry/myapp" --version "=1.2.3"
```

I know that the docker datasource doesn't support ranges, but helm does. Helm will use 1.2.3_2024-03-04.bc2e912 for =1.2.3.

In the registry, there are:

```
1.2.3-1_2024-01-17.7da0fd4
1.2.3-3_2024-01-18.5eceb63
1.2.3-4_2024-03-04.600aa2a
1.2.3_2024-03-04.600aa2a
1.2.3_2024-03-04.bc2e912
1.2.4-2_2024-03-15.a3db0f3
1.2.4-3_2024-03-15.18a741a
1.2.4_2024-03-15.18a741a
```

setting chartVersion to 1.2.3 yields 1.2.4-3_2024-03-15.18a741a, which is not the desired version.

Using packageRules, I can lock down the version the desired one:

```json
{
  "packageRules": [
    {
      "matchDepNames": [ "myapp" ],
      "matchManagers": [ "regex" ],
      "allowedVersions": "/^1.2.3(_.*)?$/"
    } 
  ]
}
```

So I'm stuck to use the workaround using HTTP(S) instead of OCI URLs, the docker datasource, and using allowedVersions in packageRules? I'd like to just use flux/helm to fulfill the versionn requirement and have renovate recommend a range.

Can this be done?

### Logs (if relevant)

<details><summary>Logs</summary>

```

Replace this text with your logs, between the starting and ending triple backticks

```

</details>

**Comment by viceice** [comment]
you need to play with versioning setting, maybe semver is appropriate for you

**Reply by fuero** [reply]
I tried semver, semver-coerced with the configuration outlined above, no dice. The docker datasource still feeds =1.2.3 to the registry and doesn't query the tags.

**Reply by viceice** [reply]
you can exclude the `=` from version match. we don't have a versioning which supports this

**Reply by fuero** [reply]
> we don't have a versioning which supports this

Ok, I outlined the workaround I use above. But as I mentioned:

> [...] doesn't support ranges, but helm does. Helm will use 1.2.3_2024-03-04.bc2e912 for =1.2.3.

Renovate's behaviour, as far as I understood, is applying the versioning module on the result it gathers from the datasource, meaning the datasource is queried **before** versioning is used, correct?

So it doesn't matter which versioning I select, as the docker datasource will treat the version as is, as evidenced [here](https://github.com/renovatebot/renovate/blob/b6d2c4bb32e52a2d8a5d7fe06b89297a9bb3d5b6/lib/modules/datasource/docker/index.ts#L121).

So to properly support helm with oci repos and ranges, the helm datasource could be changed to support OCI urls.

Adding support for using =1.2.3 to the helm versioning module should also be considered, if it's not there already.
